lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -82,91 +254,348 @@ class ConnectorSourceNotify[In: Any val]
+                  None
+                else
+                  _handler.decode(bytes)?
+                end
+
+                ifdef "trace" then
+                  @printf[I32](("Msg decoded at " + _pipeline_name +
+                    " source\n").cstring())
+                end
+
+                // get message key
+                let key_string =
+                  match m.key
+                  | None =>
+                    ""
+                  | let str: String val =>
+                    str
+                  | let a: Array[U8] val =>
+                    let k' = recover trn String(a.size()) end
+                    for c in a.values() do
+                      k'.push(c)
+                    end
+                    consume k'
+                  end
+
+                // process message
+                return _run_and_subsequent_activity(latest_metrics_id, ingest_ts,
+                  pipeline_time_spent, key_string, source, decoded, s,
+                  m.message_id, m.flags)
+              else
+                // _handler.decode(bytes) failed
+                if m.message is None then
+                  // TODO [post-source-migration] revisit this error message
+                  return _to_error_state(source, "No message bytes and BOUNDARY not set")
+                end
+                @printf[I32](("Unable to decode message at " + _pipeline_name + " source\n").cstring())
+                ifdef debug then
+                  Fail()
+                end
+                return _to_error_state(source, "Unable to decode message")
+              end
+            end
+          else
+            return _to_error_state(source, "Unknown StreamId")
           end
-          (true, ingest_ts)
         end
 
-      if is_finished then
-        let end_ts = WallClock.nanoseconds()
-        let time_spent = end_ts - ingest_ts
+      | let m: cwm.AckMsg =>
+        ifdef "trace" then
+          @printf[I32]("TRACE: got AckMsg\n".cstring())
+        end
+        return _to_error_state(source, "Invalid message: ack")
 
-        ifdef "detailed-metrics" then
-          _metrics_reporter.step_metric(_pipeline_name,
-            "Before end at Connector Source", 9999,
-            last_ts, end_ts)
+      | let m: cwm.RestartMsg =>
+        ifdef "trace" then
+          @printf[I32]("TRACE: got RestartMsg\n".cstring())
         end
+        return _to_error_state(source, "Invalid message: restart")
 
-        _metrics_reporter.pipeline_metric(_pipeline_name, time_spent +
-          pipeline_time_spent)
-        _metrics_reporter.worker_metric(_pipeline_name, time_spent)
+      end
+    else
+      @printf[I32](("Unable to decode message at " + _pipeline_name +
+        " source\n").cstring())
+      ifdef debug then
+        Fail()
+      end
+      return _to_error_state(source, "Unable to decode message")
+    end
+    _continue_perhaps(source)
+
+  fun ref _run_and_subsequent_activity(latest_metrics_id: U16,
```

Comment:
Can we rename this `_process_data_message`?

Reply:
Sure.